state/firewallrules.go

```diff
@@ -0,0 +1,178 @@
+	WhitelistCIDRS   []string `bson:"whitelist-cidrs"`
+}
+
+type firewallRules struct {
+	st  *State
+	doc firewallRulesDoc
+}
+
+// Id returns the id for the firewall rules entity.
+func (r *firewallRules) Id() string {
+	return r.doc.Id
+}
+
+// String returns r as a user-readable string.
+func (r *firewallRules) String() string {
+	return fmt.Sprintf("firewall rules for %v", r.doc.WellKnownService)
+}
+
+// WellKnownService returns the known service for the firewall rules entity.
+func (r *firewallRules) WellKnownService() WellKnownServiceType {
+	return WellKnownServiceType(r.doc.WellKnownService)
+}
+
+// BlacklistCIDRS returns the blacklist CIDRs for the rule.
+func (r *firewallRules) BlacklistCIDRS() []string {
+	return r.doc.BlacklistCIDRS
+}
+
+// WhitelistCIDRS returns the whitelist CIDRs for the rule.
+func (r *firewallRules) WhitelistCIDRS() []string {
+	return r.doc.WhitelistCIDRS
+}
+
+// FirewallRuler instances provide access to firewall rules in state.
+type FirewallRuler interface {
+	Save(service WellKnownServiceType, whiteListCidrs, blackListCidrs []string) (FirewallRules, error)
+	Rules(service WellKnownServiceType) (FirewallRules, error)
+}
+
+const (
+	// SSHRule is a rule for SSH connections.
+	SSHRule = WellKnownServiceType("ssh")
+
+	// JujuControllerRule is a rule for connections to the Juju controller.
+	JujuControllerRule = WellKnownServiceType("juju-controller")
+
+	// JujuApplicationOfferRule is a rule for connections to a Juju offer.
+	JujuApplicationOfferRule = WellKnownServiceType("juju-application-offer")
+)
+
+// WellKnownServiceType defines a service for which firewall rules may be applied.
+type WellKnownServiceType string
+
+type firewallRulesState struct {
+	st *State
+}
+
+// NewFirewallRules creates a FirewallRules instance backed by a state.
+func NewFirewallRules(st *State) *firewallRulesState {
+	return &firewallRulesState{st: st}
+}
+
+// Save stores the specified firewall rules.
+func (fw *firewallRulesState) Save(service WellKnownServiceType, whiteListCidrs, blackListCidrs []string) (FirewallRules, error) {
+	for _, cidr := range append(whiteListCidrs, blackListCidrs...) {
+		if _, _, err := net.ParseCIDR(cidr); err != nil {
+			return nil, errors.NotValidf("CIDR %q", cidr)
+		}
+	}
+	doc := firewallRulesDoc{
+		Id:               string(service),
+		WellKnownService: string(service),
+		WhitelistCIDRS:   whiteListCidrs,
+		BlacklistCIDRS:   blackListCidrs,
+	}
+	buildTxn := func(int) ([]txn.Op, error) {
+		model, err := fw.st.Model()
+		if err != nil {
+			return nil, errors.Annotate(err, "failed to load model")
+		}
+		if err := checkModelActive(fw.st); err != nil {
+			return nil, errors.Trace(err)
+		}
+
+		existing, err := fw.Rules(service)
+		if err != nil && !errors.IsNotFound(err) {
+			return nil, errors.Trace(err)
+		}
+		var ops []txn.Op
+		if err == nil {
+			ops = []txn.Op{{
+				C:      firewallRulesC,
+				Id:     existing.Id(),
+				Assert: txn.DocExists,
```

Comment:
check that they haven't changed, rather than just exist?

Reply:
The thinking here is last one wins, so just checking for existence is ok I think. It's the pattern we tend to use everywhere else.

Reply:
fair enough